### ISSUE TITLE
Expo (React Native) tracking issue

### ISSUE THREAD

**Issue by littledivy** [denoland/deno#22970]
```shell
$ deno run -A create-expo-app@latest AwesomeProject/
$ cd AwesomeProject/
$ deno task dev
```

- [x] https://github.com/denoland/deno/issues/22939
- [ ] `v8.serialize` not implemented for `node:v8`
- [ ] `expo-asset` not resolved from `node_modules/@expo`
- [ ] TypeError: Invalid URL: '/' with base '192.168.0.107:8081'
- [ ] `FsWatcher.unref` is not implemented

**Comment by millsp** [comment]
Currently fails with 

```
Task web expo start --web
Starting project at /home/pierre/Throwaway/AwesomeProject
Starting Metro Bundler
Failed to construct transformer:  TypeError: this._child.send is not a function
    at ChildProcessWorker.initialize (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/jest-worker@29.7.0/node_modules/jest-worker/build/workers/ChildProcessWorker.js:169:17)
    at new ChildProcessWorker (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/jest-worker@29.7.0/node_modules/jest-worker/build/workers/ChildProcessWorker.js:100:10)
    at WorkerPool.createWorker (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/jest-worker@29.7.0/node_modules/jest-worker/build/WorkerPool.js:30:12)
    at new BaseWorkerPool (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/jest-worker@29.7.0/node_modules/jest-worker/build/base/BaseWorkerPool.js:55:27)
    at new WorkerPool (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/jest-worker@29.7.0/node_modules/jest-worker/build/WorkerPool.js:18:1)
    at new Worker (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/jest-worker@29.7.0/node_modules/jest-worker/build/index.js:141:26)
    at FileMap._getWorker (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/metro-file-map@0.80.9/node_modules/metro-file-map/src/index.js:560:24)
    at FileMap._processFile (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/metro-file-map@0.80.9/node_modules/metro-file-map/src/index.js:382:21)
    at FileMap._applyFileDelta (file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/metro-file-map@0.80.9/node_modules/metro-file-map/src/index.js:482:33)
    at file:///home/pierre/Throwaway/AwesomeProject/node_modules/.deno/metro-file-map@0.80.9/node_modules/metro-file-map/src/index.js:214:20
error: Uncaught Error: Failed to spawn '/home/pierre/.deno/bin/deno': Argument list too long (os error 7)
```

`deno 1.44.0`
`Ubuntu 24.04 LTS`

**Comment by theopomies** [comment]
Hi, here are 2 other issues I've encountered:
```
$> deno task web
[...]
TypeError: Return value from serve handler must be a response or a promise resolving to a response
    at ext:deno_http/00_serve.ts:372:15
    at Object.runMicrotasks (ext:core/01_core.js:691:26)
    at processTicksAndRejections (ext:deno_node/_next_tick.ts:57:10)
    at runNextTicks (ext:deno_node/_next_tick.ts:75:3)
    at eventLoopTick (ext:core/01_core.js:182:21)
Exception in onError while handling exception TypeError: Return value from onError handler must be a response or a promise resolving to a response
    at ext:deno_http/00_serve.ts:381:17
    at Object.runMicrotasks (ext:core/01_core.js:691:26)
    at processTicksAndRejections (ext:deno_node/_next_tick.ts:57:10)
    at runNextTicks (ext:deno_node/_next_tick.ts:75:3)
    at eventLoopTick (ext:core/01_core.js:182:21)
Terminating Deno.serve loop due to unexpected error TypeError: Cannot read properties of undefined (reading 'status')
    at ext:deno_http/00_serve.ts:407:26
    at Object.runMicrotasks (ext:core/01_core.js:691:26)
    at processTicksAndRejections (ext:deno_node/_next_tick.ts:57:10)
    at runNextTicks (ext:deno_node/_next_tick.ts:75:3)
    at eventLoopTick (ext:core/01_core.js:182:21)
```
And then after Ctrl-C to shutdown:
```
✖ Failed to stop server
Error: Not implemented: StatWatcher.unref() is not implemented
Error: Not implemented: StatWatcher.unref() is not implemented
    at notImplemented (ext:deno_node/_utils.ts:9:9)
    at StatWatcher.unref (ext:deno_node/_fs/_fs_watch.ts:207:5)
    at FileNotifier.unsubscribe (/Users/theopomies/projects/deno/deno-expo/node_modules/.deno/@expo+cli@0.18.29/node_modules/@expo/cli/src/utils/FileNotifier.ts:78:15)
    at FileNotifier.stopObserving (/Users/theopomies/projects/deno/deno-expo/node_modules/.deno/@expo+cli@0.18.29/node_modules/@expo/cli/src/utils/FileNotifier.ts:57:5)
    at DevServerManager.stopAsync (/Users/theopomies/projects/deno/deno-expo/node_modules/.deno/@expo+cli@0.18.29/node_modules/@expo/cli/src/start/server/DevServerManager.ts:232:22)
    at KeyPressHandler.onPress (/Users/theopomies/projects/deno/deno-expo/node_modules/.deno/@expo+cli@0.18.29/node_modules/@expo/cli/src/start/interface/startInterface.ts:75:34)
    at ReadStream.KeyPressHandler.handleKeypress (/Users/theopomies/projects/deno/deno-expo/node_modules/.deno/@expo+cli@0.18.29/node_modules/@expo/cli/src/start/interface/KeyPressHandler.ts:43:18)
    at ReadStream.emit (ext:deno_node/_stream.mjs:1851:9)
    at addChunk (ext:deno_node/_stream.mjs:2873:16)
    at readableAddChunk (ext:deno_node/_stream.mjs:2847:15)
```
```
deno 2.0.0-rc.1 (release candidate, release, aarch64-apple-darwin)
v8 12.9.202.5-rusty
typescript 5.5.2
```
` "expo": "~51.0.28"`

Repro is fairly easy:
```sh
deno run -A npm:create-expo-app@latest deno-expo
cd deno-expo
deno task web
```

**Comment by sigmaSd** [comment]
Seems like some upstream patch is also needed because currently running expo via deno, runs bunx instead

![image](https://github.com/user-attachments/assets/c3192c14-ca8a-41d7-8bca-0dc3b08a0ea5)

if I hide bun it run npx

![image](https://github.com/user-attachments/assets/c2a34899-799c-4add-a225-856b4b47e042)

**Comment by sigmaSd** [comment]
Also if I follow divy steps and create the project and add deno.json (with or without its the same error), deno task start still fails
```
TypeError: Cannot read properties of undefined (reading 'headerList')
```

**Comment by sleepy-goth** [comment]
I tried on an already configured project to run "deno install" and the process became cpu intesive without output, had to terminate.

**Comment by sigmaSd** [comment]
The deno install hanging is fixed in deno 2.2.6, but the main issue remains it can't run the app it failswith the same error undefined reading headerlist

**Comment by knotbin** [comment]
@sigmaSd is there an issue tracking that error? Is it still an issue?

**Comment by robertjpayne** [comment]
~~With deno v2.3.7 on macOS running `deno install` at the root of a expo-app template deno basically runs until it consumes all ram and fails due to memory exhaustion.~~

Updated: turns out this was just the type checking Deno does on install, since expo is very large it takes a very long time and I was killing it before it completed.

**Comment by knotbin** [comment]
> Also if I follow divy steps and create the project and add deno.json (with or without its the same error), deno task start still fails
> 
> ```TypeError: Cannot read properties of undefined (reading 'headerList')```

@sigmaSd @littledivy through a ton of debug logging I did about a week ago (don't have exact logs anymore sorry), I found that expo is sending 2 requests to deno's fetch impl, one that is a proper request with a URL, headers, referrer, etc. and one that is only one property iirc a blank headers. The reason it was so hard to debug this is that I would get tripped up by the first request and assume that request was causing the error. Don't make that mistake if you want to try and debug this yourself. 

This error is not technically Deno's fault but Deno shouldn't error out when it did and should actually error out earlier, when constructing the request object in request.js. I assume node/undici's implementation of fetch simply ignores the invalid request.

**The easy solution**

In the expo repo, specifically packages/@expo/cli/src/api/src/utils/fetch.ts:

```ts
export const fetch: FetchLike =
  typeof globalThis.fetch !== 'undefined' ? globalThis.fetch : require('undici').fetch;

export const Headers: typeof import('undici').Headers =
  typeof globalThis.Headers !== 'undefined' ? globalThis.Headers : require('undici').Headers;
```

When I change this to simply use `require('undici').Headers` and `require('undici').fetch`, `deno task dev` works fine because undici has a more permissive fetch function. Theoretically one could open a pull request in expo adding deno overrides to their fetch function but I honestly doubt they would accept such a PR.

**The real solution**

What I haven't been able to do is find where in packages/@expo/cli/src/api/src/rest (I'm 99% sure it's in that folder) the _second_ request actually happens and because I'm not an actual user of Expo currently for work or personal projects, I can't really spend any more time on this atm but I really do hope someone picks this up or that I can come back to it at some point because I've confirmed running locally that once this last piece works, expo actually **RUNS**.

**Comment by sigmaSd** [comment]
It seems expo relies on HeaderList being present, It seems to be specific to undici and not part of the web spec

Nice find though, atleast now there is a workaround
```ts
deno --preload <(echo "globalThis.fetch = undefined; globalThis.Headers = undefined;") -A node_modules/@expo/cli/build/bin/cli #bash
```

Though I noticed that deno install seems to hang sometimes, it seems like if I downgrade to 2.2.6 it works more reliably

**Comment by sigmaSd** [comment]
Actually even with that patch , its true it doesn't error, but if I try to connect my phone to it it just loads foever, and if I test with bun it works, I think the workaround is still not enough

**Comment by knotbin** [comment]
I got errors when connecting on iOS but not loading forever and for me it worked to open on web. 

Of course there's a ton more work to do for it to fully work and we're mostly taking baby steps but I feel like this is a huge barrier broken for this issue after over a year.

All we can do is just keep working through the logs towards full compat.  It'll feel great when it finally gets there.

**Comment by bartlomieju** [comment]
Re-tested with Deno 2.8.0 on macOS (aarch64), scaffolding from scratch:

```
deno run -A npm:create-expo-app@latest deno-expo --template blank
cd deno-expo
deno install
deno add npm:react-dom npm:react-native-web    # blank template doesn't include these
deno task web
```

The web flow works end-to-end:

- `deno install` finishes in ~1s, no hang.
- Metro starts and reports `Waiting on http://localhost:8081`.
- `GET /` returns the Expo HTML shell.
- `GET /index.bundle?platform=web&...` returns a 1.5 MB bundle (211 modules) in ~2.5s.
- Ctrl-C shuts the server down cleanly (exit 0).

None of the symptoms reported earlier in this thread reproduce on current Deno:

- `this._child.send is not a function` (jest-worker)
- `Argument list too long` when spawning deno
- `Return value from serve handler must be a response`
- `Cannot read properties of undefined (reading 'headerList')`
- `StatWatcher.unref() is not implemented` on shutdown
- `deno install` hang

I did not test the iOS/Android device path (no phone/simulator on hand), which is the remaining concern from the most recent comments. If that path is still broken on current Deno, it'd be more actionable as a fresh, narrower issue than this 2-year-old umbrella. Closing for now — please open a focused follow-up if device connections still hang.
